google/cloud/aiplatform/prediction/handler.py

```diff
@@ -103,15 +108,56 @@ async def handle(self, request: Request) -> Response:
 
         Returns:
             The response of the prediction request.
+
+        Raises:
+            HTTPException: If any expection is thrown from deserialization or serialization.
         """
         request_body = await request.body()
-        content_type = handler_utils.get_content_type_from_headers(request.headers)
-        prediction_input = DefaultSerializer.deserialize(request_body, content_type)
+        try:
+            content_type = handler_utils.get_content_type_from_headers(request.headers)
+            prediction_input = DefaultSerializer.deserialize(request_body, content_type)
+        except HTTPException:
+            raise
+        except Exception as exception:
+            error_message = "An exception {} occurred. Arguments: {}.".format(
+                type(exception).__name__, exception.args
+            )
+            logging.info(
+                "{}\\nTraceback: {}".format(error_message, traceback.format_exc())
+            )
+
+            # Converts all other exceptions to HTTPException.
+            raise HTTPException(
+                status_code=500,
+                detail=error_message,
+                headers={
+                    prediction.CUSTOM_PREDICTION_ROUTINES_SERVER_ERROR_HEADER_KEY: version.__version__
+                },
+            )
 
```

Comment:
Initial thought is that we should wrap the user's code in try, instead of trying to cover all the places our code is executed. Still thinking about the best way to do that.

Reply:
If I understand correctly, what your thought will look like:
handler:
```
try:
  ...user code...
except:
  raises HTTPException without headers
```
model server:
```
def predict:
  try:
    return self.handler.handle(...)
  except HTTPException:
    raise (without headers)
  except Exception:
    raise with headers
```
The issue is that we allow custom handlers which users may not raise all potential exceptions in their code. That's why I did a reverse way.